### PR TITLE
Pull request for merging get_prime_numbers function to main

### PR DESCRIPTION
**Pull request by Melat-arch**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_get_prime_numbers.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# tests/test_prime_numbers.py
+
+"""
+Unit tests for the get_prime_numbers function.
```
Comment (Idris-Pamiri) [inline solutions/tests/test_get_prime_numbers.py:4]: Would be good if types of unit tests are described right int he module unit test. 
Like what type of test are you running; regular, edge cases, etc>

File: solutions/tests/test_get_prime_numbers.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_primes_up_to_10(self):
+        """Test that the function returns correct primes up to 10."""
+        self.assertEqual(get_prime_numbers(10), [2, 3, 5, 7])
+
+    def test_primes_up_to_2(self):
+        """Test that the function returns correct primes up to 2."""
+        self.assertEqual(get_prime_numbers(2), [2])
+
+    def test_no_primes_below_2(self):
```
Comment (Idris-Pamiri) [inline solutions/tests/test_get_prime_numbers.py:24]: The checklist provided by Evan requires us to have only 1 assertion per Unit Test. The reason is is done (only 1 per unit-test) is to make sure that when you debug you know exactly which unit test failed. Otherwise, you 'll have to guess which one of these 3 caused your unit test to fail. 
As a option, you can split these into 3 different test cases (in one of the Q&A Videos Evan explains it as well)

File: solutions/tests/test_get_prime_numbers.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        """Test that the function returns an empty list for n < 2."""
+        self.assertEqual(get_prime_numbers(1), [])
+        self.assertEqual(get_prime_numbers(0), [])
+        self.assertEqual(get_prime_numbers(-5), [])
+
+    def test_large_number(self):
+        """Test the function with a larger number."""
+        self.assertEqual(get_prime_numbers(20), [2, 3, 5, 7, 11, 13, 17, 19])
+
```
Comment (Idris-Pamiri) [inline solutions/tests/test_get_prime_numbers.py:33]: As part of the defensive strategy, i would suggest to add a unit test that will raise assertion when a non integer (say string or a character or a symbol) is passed as an argument to the function. That will have to be written bot in the test file and also in the function itself once it "detects" a non-integer argument.

File: solutions/get_prime_numbers.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    Examples:
+        >>> get_prime_numbers(10)
+        [2, 3, 5, 7]
+        >>> get_prime_numbers(2)
+        [2]
+        >>> get_prime_numbers(1)
+        []
+    """
+    if n < 2:
```
Comment (Idris-Pamiri) [inline solutions/get_prime_numbers.py:36]: Maybe as part of the behavior, try to do defensive assertion for the mismatch in the type of the argument - if a non-integer is passed to your function - so that it raises an assertion in that case. (also flagged it in the unit-test file)

**Review by Idris-Pamiri — CHANGES_REQUESTED** [review]
Dear Melat.
Great work on the function. Overall function and the tests work great. I provided some inputs regarding the behavior and strategy  - once you look at those and make appropriate changes, the files will be good to go with the merging. 
Great job.

File: solutions/tests/test_get_prime_numbers.py (hunk 3 above)
Comment (Melat-arch) [reply solutions/tests/test_get_prime_numbers.py:33]: Thank you for your feedback. But I had a question. If the question explicitly states that the input will only contain integers, isn't it reasonable to mainly focus on test cases that align with the input type specified in the question?

File: solutions/tests/test_get_prime_numbers.py (hunk 1 above)
Comment (Melat-arch) [reply solutions/tests/test_get_prime_numbers.py:4]: That is a very good suggestion; I will definitely apply it. Thank you.

File: solutions/tests/test_get_prime_numbers.py (hunk 3 above)
Comment (Idris-Pamiri) [reply solutions/tests/test_get_prime_numbers.py:33]: @Melat-arch correct. We can leave it as per the wording used in the text of the challenge. My comment was more around the fact that we have the defensive coding "hat" on expecting some non-integer input. Just to make sure that that error is being handled if it "appears".

**Comment by Idris-Pamiri** [comment]
@Melat-arch - as you saw in the group chat, due to the updated extension of the ruff extension there is an error in the CI pipeline check. I suggest you re-check your code in VS code and create a new pull request and link it ..that way we can push it once again. This time, given the discussion above I can reivew and merge it right away.
